Fix for WFLY-14411, Server components upgrade and Galleon feature-pack dependency update

jfdenise opened the pull request:
Analysis that describes the feature: https://github.com/wildfly/wildfly-proposals/pull/372

NB: All UpdateArtifact*TestCase are meaningless until we test using WF23 Beta1.

jfdenise commented:
The 2 failing tests UpgradeArtifactFPLTestCase and UpgradeArtifactTestCase are expected to fail.

jamezp reviewed (commented):
All the comments are pretty minor or just questions.

Does this require a provisioning upgrade?

jfdenise commented on plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/MavenUpgrade.java:
```diff
@@ -0,0 +1,302 @@
+import org.wildfly.plugins.bootablejar.maven.common.FeaturePack;
+import org.wildfly.plugins.bootablejar.maven.common.OverriddenArtifact;
+
+public final class MavenUpgrade {
```
Thanks fixed.

jfdenise commented on plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/MavenProjectArtifactVersions.java:
```diff
@@ -0,0 +1,75 @@
+import org.apache.maven.project.MavenProject;
+import org.wildfly.plugins.bootablejar.maven.common.OverriddenArtifact;
+
+public final class MavenProjectArtifactVersions {
```
Thanks fixed.

jfdenise commented on plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/MavenUpgrade.java:
```diff
@@ -0,0 +1,302 @@
+            resolvedFeaturePacks.put(fp.getGAC(), mojo.resolveMaven(fp));
+        }
+        mojo.getLog().debug("Top level feature-packs: " + topLevels);
+        mojo.getLog().debug("Resolved feature-packs " + resolvedFeaturePacks);
```
Thanks, fixed.

jfdenise commented on plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/MavenUpgrade.java:
```diff
@@ -0,0 +1,302 @@
+                mojo.getLog().info("[UPDATE] Overriding Galleon feature-pack dependency with: ");
+                for (FeaturePack fp : featurePackDependencies) {
+                    FeaturePackLocation fpl = FeaturePackLocation.fromString(fp.getMavenCoords());
+                    mojo.getLog().info("[UPDATE]  " + fp.getGroupId() + ":" + fp.getArtifactId() + ":"
```
Very good point, I added version check and warn for downgrade.

jfdenise replied:
We already have a toString defined on feature-pack that contains much more. I didn't change anything there, I could introduce a new method to handle this formatting.

jamezp replied:
Ah perfect. Makes sense then.

jfdenise commented:
@jamezp , I made quite a lot of changes. Implemented your feedbacks. What do you mean by "provisioning upgrade"?. The galleon-plugins 5.1.0 (contained by WF feature-pack), starting WF23, will contain what is needed for the upgrade of artifacts to work.

 In addition I added:
* Ability to dump all the artifacts a user can upgrade. This seems required to understand what can be done.
* Added community doc.
* I reduced the scope of known artifacts to the provisioned ones. It is required now that a list of artifacts can be dumped. All artifacts would be useless. User needs to know the actual set of artifacts present in the bootable JAR.

jfdenise commented on plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/MavenUpgrade.java:
```diff
@@ -0,0 +1,302 @@
+                    throw new MojoExecutionException("No version for Galleon feature-pack " + a.getGAC());
+                } else {
+                    FeaturePack dep = dependencies.get(key);
+                    if (fpArtifact.getVersion().equals(dep.getVersion())) {
```
That is a very good point. I added version check and warn for downgrade. Analysis reflect that we can downgrade.

jamezp commented:
Sorry. Yeah I meant is there a galleon upgrade required. Not sure why I said provisioning :)

jfdenise commented on plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/MavenUpgrade.java:
```diff
@@ -0,0 +1,413 @@
+                } else {
+                    FeaturePack dep = dependencies.get(key);
+                    MavenArtifactVersion orig = new MavenArtifactVersion(dep.getVersion());
+                    MavenArtifactVersion overridenVersion = new MavenArtifactVersion(fpArtifact.getVersion());
```
Thank-you!

jfdenise commented on plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/MavenUpgrade.java:
```diff
@@ -0,0 +1,413 @@
+                        throw new MojoExecutionException("Overridden artifact " + a.getGAC() + " not know in provisioned feature-packs");
+                    }
+                    MavenArtifactVersion orig = new MavenArtifactVersion(originalVersion);
+                    MavenArtifactVersion overridenVersion = new MavenArtifactVersion(a.getVersion());
```
Thank-you!

jfdenise commented:
@jamezp , thanks for your review, merged.
